### PR TITLE
use optionalt<exprt> instead of nil_exprt in expr_initializer

### PR DESCRIPTION
**Pull request by kroening**
This matches the signature of the external interface.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c2ec056adef6c98abcd7d661cea838f4df2d058d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99984282

File: src/util/expr_initializer.cpp [hunk 1]
```diff
@@ -241,62 +240,78 @@ exprt expr_initializert<nondet>::expr_initializer_rec(
...
+    if(!result.has_value())
+      return {};
+
     // we might have mangled the type for arrays, so keep that
     if(type.id() != ID_array)
-      result.type()=type;
+      result->type() = type;
 
-    return result;
+    return *result;
```
Comment (tautschnig) [inline src/util/expr_initializer.cpp:275]: `operator*` actually isn't necessary.

File: src/util/expr_initializer.cpp [hunk 2]
```diff
@@ -241,62 +240,78 @@ exprt expr_initializert<nondet>::expr_initializer_rec(
...
 
+    if(!result.has_value())
+      return {};
+
     // use the tag type
-    result.type() = type;
+    result->type() = type;
 
-    return result;
+    return *result;
```
Comment (tautschnig) [inline src/util/expr_initializer.cpp:288]: `operator*` actually isn't necessary.

File: src/util/expr_initializer.cpp [hunk 3]
```diff
@@ -241,62 +240,78 @@ exprt expr_initializert<nondet>::expr_initializer_rec(
...
```
Comment (tautschnig) [inline src/util/expr_initializer.cpp:301]: `operator*` actually isn't necessary.

File: src/util/expr_initializer.cpp [hunk 4]
```diff
@@ -241,62 +240,78 @@ exprt expr_initializert<nondet>::expr_initializer_rec(
...
```
Comment (tautschnig) [inline src/util/expr_initializer.cpp:314]: `operator*` actually isn't necessary.

**Review by tautschnig — APPROVED** [review]

**Comment by kroening** [comment]
On the `operator*`: I did think about this one quite a bit before adding the `*`.
My rationale was that this makes it clear that we return at that location with a non-error case.

**Pull request merged**